### DISCUSSION TITLE
clean jpeg noise

### DISCUSSION THREAD

**Discussion by exekutive** [ImageMagick/ImageMagick#7442] [Help] [answered]
### ImageMagick version

7.1.1

### Operating system, version and so on

darwin

### Description

What is the most effective way to remove noise from grayscale text (leftover jpeg compression artifacts)?

examples:

![noise1](https://github.com/ImageMagick/ImageMagick/assets/4755215/1c05a836-44b2-4f1f-99c9-d5daf2ddae8c)
![noise2](https://github.com/ImageMagick/ImageMagick/assets/4755215/a5ffdd4e-7691-40b2-9a0a-5c7cf55886ff)

**Comment by exekutive** [comment]
Also, the images were enlarged 400%. Is it better to de-noise before enlarging, or does it not matter?

**Comment by fmw42** [comment]
I would say try before enlarging at the original resolution. Can you post the original for that file?

**Comment by exekutive** [comment]
here they are before enlarging

![header2_](https://github.com/ImageMagick/ImageMagick/assets/4755215/7370e3e3-a251-44e0-8e1d-d3e041082e40)
![header3_](https://github.com/ImageMagick/ImageMagick/assets/4755215/3c2c7df6-858a-4fce-9fcf-e70b98afd3da)

**Reply by fmw42** [reply]
Why do these jpg's need cleaning?  What do you need to do with these images if they are not acceptable?  That is what is your end goal for these image? Is it just to zoom them?

**Reply by exekutive** [reply]
The immediate goal is to get the images reasonably clean before I run OCR on them.
The long term goal is to learn how to use imagemagick and general techniques. I can see myself using more programmatic image manipulation in the future.
De-noising and removing artifacts seems like a useful way to improve the quality of images , wouldn't you say?

**Comment by fmw42** [comment]
It is hard to really remove JPG compression noise. Imagemagick does not have anything specific for the task especially for JPG noise. For OCR; I would suggest just thresholding with -auto-threshold otsu in IM 7 and save to PNG (not JPG).  P.S. the files you posted are PNGs not JPGS. So I do not know what else you have tried on your original JPGS of which I asked you to post. But if you only converted them to PNG and did nothing else, then no problem.

Here are my results

```
magick image1.png -auto-threshold otsu x1.png
magick image2.png -auto-threshold otsu x2.png
```


![x1](https://github.com/ImageMagick/ImageMagick/assets/13458218/8e6a303f-10fa-4cde-8ffa-cf0b19733ec4)

![x2](https://github.com/ImageMagick/ImageMagick/assets/13458218/a20da16d-17bd-443b-b516-0027bf40c7c5)

**Reply by fmw42** [reply]
P.S. I suggest you use larger font sizes for OCR.  It will have trouble if you fonts are too small.

You could also try wavelet denoising and see if that helps.  See https://imagemagick.org/script/command-line-options.php#wavelet-denoise

**Reply by fmw42** [reply]
Zooming the images you posted, I see no noise!  Are you sure you posted the original JPG images?

**Reply by exekutive** [reply]
>  I would suggest just thresholding with -auto-threshold otsu

Tesseract does otsu internally. 

>  I suggest you use larger font sizes for OCR.

That's why they are enlarged 400%

> Zooming the images you posted, I see no noise! Are you sure you posted the original JPG images?

it's definitely there. Are you zooming at the thresholded ones you posted?

**Reply by exekutive** [reply]
<img width="1090" alt="Screenshot 2024-06-30 at 13 52 50" src="https://github.com/ImageMagick/ImageMagick/assets/4755215/520bd2a2-8a34-4107-9f6d-c8852efddac9">

**Reply by fmw42** [reply]
Yes, sorry, I zoomed the wrong images - mine after threshold.  If OCR does thresholding, then I suppose you do not have to worry about it.  Did -despeckle or -enhance help?  How about wavelet denoising?

Did you do any research on Google for JPG denoising?

**Reply by fmw42** [reply]
See topaz at https://www.topazlabs.com/denoise-ai?g_campaign=S-PROS-US_CA-DSA&g_acctid=883-417-5396&g_adgroupid=145930533531&g_keyword=&g_campaignid=18076184622&g_adid=616953141105&g_network=g&g_adtype=search&g_keywordid=dsa-1497421536377&gad_source=1&gclid=EAIaIQobChMI1rbq7qGEhwMVdhmtBh1eiQFuEAMYAiAAEgIAFfD_BwE
for generative AI solution.

**Reply by exekutive** [reply]
enhance is pretty good but not perfect. I still need to experiment with wavelet and compare.
Maybe I don't need to worry, but I'd still like to improve the accuracy as much as possible, and it will be a useful to know how to do.
I'm trying to stick to basic CLI tools like IM which are ubiquitous and fast. I don't want to touch AI at all.

**Comment by fmw42** [comment]
Try https://ai.nero.com/denoiser. It seems to work on your first image (only one I tried)

**Reply by exekutive** [reply]
not interested in online tools. IM only. I _might_ look into VIPS

**Comment by fmw42** [comment]
try 

`magick image1.png -kuwahara 0.1 x1.png`

![x1](https://github.com/ImageMagick/ImageMagick/assets/13458218/1b71f44a-81b2-4951-8a12-2afae1c0e3d5)

It's a little fuzzy, but the noise has been mitigated. Adjust the argument as desired.

**Reply by exekutive** [reply]
first impression is that noise is still there, just fuzzier, and I'd say that overall image fidelity is worse than before

**Comment by exekutive** [comment]
using the original sizes, I think -despeckle actually wins.

Here are the various results ....

image 1: `-despeckle -despeckle -despeckle `

- A little bit of noise remains immediately adjacent the text.
- Overall contrast was mostly maintained.
- Going beyond 3x degrades the image too much. Black begins to fade to grey, and the white inside of the letters start to fill with grey.

![head2~speck3~web](https://github.com/ImageMagick/ImageMagick/assets/4755215/4c2d395c-bf24-43d4-8cc8-cfa268bab78b)

image 2: `-enhance -enhance -enhance -enhance -enhance`

- Significant amount of noise left untouched.
- contrast and sharpness is very well maintained
- the effectiveness of `-enhance` diminishes very quickly. After about 3x, there is almost no visible change, however blacks & whites are well preserved.

![head2~enh5~web](https://github.com/ImageMagick/ImageMagick/assets/4755215/2000ebca-84fb-446a-9e3a-5c32b4a46445)


image 3: `-wavelet-denoise 10%`
usage: `-wavelet-denoise thresholdxsoftness`

- wavelet removes noise from the open areas but not adjacent the text (opposite of what we want)
- beyond 10% , black text quickly turns grey, blurry, and halos form around letters, white turns to grey, 
- I did not have any success at all with the softness parameter. Using anything but zero made the image binary and turned the noise dark black

![head2~wave10~web](https://github.com/ImageMagick/ImageMagick/assets/4755215/a262c5e5-13af-4ced-8938-b668b10e503e)


image 4: `-kuwahara 0.1`

- kuwahara was the worst of them all
- claims to be "Edge preserving noise reduction filter."  but it's not what I see at all
- badly blurs the text at any value. Beyond a value of '1' text becomes completely unreadable
- unusable for text in my opinion

![head2~kuwa01~web](https://github.com/ImageMagick/ImageMagick/assets/4755215/846f214d-feb8-49d3-9529-7b8a3dc99cb4)

**Reply by fmw42** [reply]
try -mean-shift 8x8+30%.  It seemed to work well for the first image, not so well on the second.

**Reply by exekutive** [reply]
Hey... that's really good!

![mean](https://github.com/ImageMagick/ImageMagick/assets/4755215/13a76fc9-d890-41b6-a046-616ed3fc4aa8)

**Reply by exekutive** [reply, marked as answer]
second image. Still an improvement
![head3~crop](https://github.com/ImageMagick/ImageMagick/assets/4755215/f6a9bf46-79b3-4699-832b-8386c8b77579)
![head3~mean](https://github.com/ImageMagick/ImageMagick/assets/4755215/b929a0d7-8a5e-4f7c-b59c-8c072cc69700)

**Comment by fmw42** [comment]
You can try to adjust the arguments further and see if that helps.  I guessed at 8x8 since that is the size of the JPG compression, but other values may be better.

**Reply by exekutive** [reply]
you're right. with anti-aliased text (like image 2 with some gray pixels), mean 8x8 erodes the letters too much. A value like 4x4+30 is safer. It still cleans most of the noise without too much erosion. or even 4x4+20. Beyond 30% there is too much blurring.
